examples/lock-app/silabs/src/LockManager.cpp

```diff
@@ -353,34 +333,103 @@ bool LockManager::Unbolt(chip::EndpointId endpointId, const Nullable<chip::Fabri
 
 bool LockManager::GetUser(chip::EndpointId endpointId, uint16_t userIndex, EmberAfPluginDoorLockUserInfo & user)
 {
+    CHIP_ERROR error;
+
     VerifyOrReturnValue(userIndex > 0, false); // indices are one-indexed
 
     userIndex--;
 
     VerifyOrReturnValue(IsValidUserIndex(userIndex), false);
 
+    VerifyOrReturnValue(kInvalidEndpointId != endpointId, false);
+
     ChipLogProgress(Zcl, "Door Lock App: LockManager::GetUser [endpoint=%d,userIndex=%hu]", endpointId, userIndex);
 
-    const auto & userInDb = mLockUsers[userIndex];
+    // Get User struct from nvm3
+    chip::StorageKeyName userKey = LockUserEndpoint(endpointId, userIndex);
 
-    user.userStatus = userInDb.userStatus;
-    if (UserStatusEnum::kAvailable == user.userStatus)
+    uint16_t size = LockUserInfoSize;
+
+    LockUserInfo userInStorage;
+
+    error = mStorage->SyncGetKeyValue(userKey.KeyName(), &userInStorage, size);
+
+    // Check size out param matches what we expect to read
+    VerifyOrReturnValue(size == LockUserInfoSize, false);
+
+    // If no data is found at user key
+    if (error == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
     {
-        ChipLogDetail(Zcl, "Found unoccupied user [endpoint=%d]", endpointId);
+        user.userStatus = UserStatusEnum::kAvailable;
+
+        ChipLogDetail(Zcl, "No user data found");
         return true;
     }
+    // Else if KVS read was successful
+    else if (error == CHIP_NO_ERROR)
+    {
+        user.userStatus = userInStorage.userStatus;
+
+        if (userInStorage.userStatus == UserStatusEnum::kAvailable)
+        {
+            ChipLogDetail(Zcl, "Found unoccupied user [endpoint=%d]", endpointId);
+            return true;
+        }
+    }
+    else
+    {
+        ChipLogError(Zcl, "Error reading from KVS key");
+        return false;
+    }
+
+    VerifyOrReturnValue(userInStorage.currentCredentialCount <= kMaxCredentialsPerUser, false);
+
+    VerifyOrReturnValue(userInStorage.userNameSize <= DOOR_LOCK_MAX_USER_NAME_SIZE, false);
```

Comment:
```suggestion
    VerifyOrReturnValue(userInStorage.userNameSize <= user.userName.size(), false);
```
since that's the actual invariant we are relying on, right?